### ISSUE TITLE
Error on installing braze sdk in react native app

### ISSUE THREAD

**Issue by vitorreis** [Appboy/appboy-android-sdk#106]
* [x] I have checked [Braze's documentation](https://www.braze.com/documentation/Android/) and am still experiencing the issue.

# Report

## What version of the Braze Android SDK are you using?
Latest -> 2.6.0

## What did you do?
I am trying to install braze SDK in a react-native app

## What did you expect to happen?
`npm run android` should start the app without errors 

## What happened instead?

```
Building and installing the app on the device (cd android && ./gradlew installDebug)...
Incremental java compilation is an incubating feature.
:app:preBuild UP-TO-DATE
:app:preDebugBuild UP-TO-DATE
:app:checkDebugManifest
:app:preReleaseBuild UP-TO-DATE
:react-native-appboy-sdk:preBuild UP-TO-DATE
:react-native-appboy-sdk:preReleaseBuild UP-TO-DATE
:react-native-appboy-sdk:checkReleaseManifest
:react-native-appboy-sdk:preDebugAndroidTestBuild UP-TO-DATE
:react-native-appboy-sdk:preDebugBuild UP-TO-DATE
:react-native-appboy-sdk:preDebugUnitTestBuild UP-TO-DATE
:react-native-appboy-sdk:preReleaseUnitTestBuild UP-TO-DATE
:react-native-appboy-sdk:prepareAndroidArchCoreRuntime110Library
:react-native-appboy-sdk:prepareAndroidArchLifecycleLivedataCore110Library
:react-native-appboy-sdk:prepareAndroidArchLifecycleRuntime110Library
:react-native-appboy-sdk:prepareAndroidArchLifecycleViewmodel110Library
:react-native-appboy-sdk:prepareComAndroidSupportAppcompatV72301Library
:react-native-appboy-sdk:prepareComAndroidSupportRecyclerviewV72301Library
:react-native-appboy-sdk:prepareComAndroidSupportSupportCompat2710Library
:react-native-appboy-sdk:prepareComAndroidSupportSupportCoreUi2710Library
:react-native-appboy-sdk:prepareComAndroidSupportSupportCoreUtils2710Library
:react-native-appboy-sdk:prepareComAndroidSupportSupportFragment2710Library
:react-native-appboy-sdk:prepareComAndroidSupportSupportV42301Library
:react-native-appboy-sdk:prepareComAppboyAndroidSdkBase240Library
:react-native-appboy-sdk:prepareComAppboyAndroidSdkUi240Library
:react-native-appboy-sdk:prepareComFacebookFrescoDrawee081Library
:react-native-appboy-sdk:prepareComFacebookFrescoFbcore081Library
:react-native-appboy-sdk:prepareComFacebookFrescoFresco081Library
:react-native-appboy-sdk:prepareComFacebookFrescoImagepipeline081Library
:react-native-appboy-sdk:prepareComFacebookFrescoImagepipelineOkhttp081Library
:react-native-appboy-sdk:prepareComFacebookReactReactNative0191Library
:react-native-appboy-sdk:prepareOrgWebkitAndroidJscR174650Library
:react-native-appboy-sdk:prepareReleaseDependencies
:react-native-appboy-sdk:compileReleaseAidl
:react-native-appboy-sdk:compileReleaseNdk UP-TO-DATE
:react-native-appboy-sdk:compileLint
:react-native-appboy-sdk:copyReleaseLint UP-TO-DATE
:react-native-appboy-sdk:compileReleaseRenderscript
:react-native-appboy-sdk:generateReleaseBuildConfig
:react-native-appboy-sdk:generateReleaseResValues
:react-native-appboy-sdk:generateReleaseResources
:react-native-appboy-sdk:mergeReleaseResources
:react-native-appboy-sdk:processReleaseManifest
:react-native-appboy-sdk:processReleaseResources
warning: string 'catalyst_debugjs' has no default translation.
warning: string 'catalyst_element_inspector' has no default translation.
warning: string 'catalyst_jsload_error' has no default translation.
warning: string 'catalyst_jsload_message' has no default translation.
warning: string 'catalyst_jsload_title' has no default translation.
warning: string 'catalyst_reloadjs' has no default translation.
warning: string 'catalyst_settings' has no default translation.
warning: string 'catalyst_settings_title' has no default translation.

:react-native-appboy-sdk:generateReleaseSources
:react-native-appboy-sdk:incrementalReleaseJavaCompilationSafeguard
:react-native-appboy-sdk:compileReleaseJavaWithJavac
:react-native-appboy-sdk:compileReleaseJavaWithJavac - is not incremental (e.g. outputs have changed, no previous execution, etc.).
:react-native-appboy-sdk:extractReleaseAnnotations
:react-native-appboy-sdk:mergeReleaseShaders
:react-native-appboy-sdk:compileReleaseShaders
:react-native-appboy-sdk:generateReleaseAssets
:react-native-appboy-sdk:mergeReleaseAssets
:react-native-appboy-sdk:mergeReleaseProguardFiles
:react-native-appboy-sdk:packageReleaseRenderscript UP-TO-DATE
:react-native-appboy-sdk:packageReleaseResources
:react-native-appboy-sdk:processReleaseJavaRes UP-TO-DATE
:react-native-appboy-sdk:transformResourcesWithMergeJavaResForRelease
:react-native-appboy-sdk:transformClassesAndResourcesWithSyncLibJarsForRelease
:react-native-appboy-sdk:mergeReleaseJniLibFolders
:react-native-appboy-sdk:transformNative_libsWithMergeJniLibsForRelease
:react-native-appboy-sdk:transformNative_libsWithSyncJniLibsForRelease
:react-native-appboy-sdk:bundleRelease
:app:prepareAndroidArchCoreRuntime110Library
:app:prepareAndroidArchLifecycleLivedataCore110Library
:app:prepareAndroidArchLifecycleRuntime110Library
:app:prepareAndroidArchLifecycleViewmodel110Library
:app:prepareAppWithBrazeSDKReactNativeAppboySdkUnspecifiedLibrary
:app:prepareComAndroidSupportAppcompatV72301Library UP-TO-DATE
:app:prepareComAndroidSupportSupportCompat2710Library
:app:prepareComAndroidSupportSupportCoreUi2710Library
:app:prepareComAndroidSupportSupportCoreUtils2710Library
:app:prepareComAndroidSupportSupportFragment2710Library
:app:prepareComAndroidSupportSupportV42301Library UP-TO-DATE
:app:prepareComAppboyAndroidSdkBase260Library
:app:prepareComAppboyAndroidSdkUi260Library
:app:prepareComFacebookFbuiTextlayoutbuilderTextlayoutbuilder100Library UP-TO-DATE
:app:prepareComFacebookFrescoDrawee130Library UP-TO-DATE
:app:prepareComFacebookFrescoFbcore130Library UP-TO-DATE
:app:prepareComFacebookFrescoFresco130Library UP-TO-DATE
:app:prepareComFacebookFrescoImagepipeline130Library UP-TO-DATE
:app:prepareComFacebookFrescoImagepipelineBase130Library UP-TO-DATE
:app:prepareComFacebookFrescoImagepipelineOkhttp3130Library UP-TO-DATE
:app:prepareComFacebookReactReactNative0554Library UP-TO-DATE
:app:prepareComFacebookSoloaderSoloader010Library UP-TO-DATE
:app:prepareOrgWebkitAndroidJscR174650Library UP-TO-DATE
:app:prepareDebugDependencies
:app:compileDebugAidl
:app:compileDebugRenderscript UP-TO-DATE
:app:generateDebugBuildConfig UP-TO-DATE
:app:mergeDebugShaders UP-TO-DATE
:app:compileDebugShaders UP-TO-DATE
:app:generateDebugAssets UP-TO-DATE
:app:mergeDebugAssets
:app:generateDebugResValues UP-TO-DATE
:app:generateDebugResources UP-TO-DATE
:app:mergeDebugResources
:app:bundleDebugJsAndAssets SKIPPED
:app:processDebugManifest
:app:processDebugResources
ERROR: In <declare-styleable> FontFamilyFont, unable to find attribute android:font
ERROR: In <declare-styleable> FontFamilyFont, unable to find attribute android:fontStyle
ERROR: In <declare-styleable> FontFamilyFont, unable to find attribute android:fontWeight

:app:processDebugResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:processDebugResources'.
> com.android.ide.common.process.ProcessException: Failed to execute aapt
```

## Steps to reproduce

- Clone the example project
- Add a `android/local.properties` file with the content `sdk.dir=/my-path/Library/Android/sdk`
- `npm run android` or open th android studio project and sync gradle

## Are you doing any feature customizations that may relate to the issue? Can you share the code snippet?
no

## Project that demonstrates the issue
https://github.com/vitorreis/react-native-braze-sdk-app

Commit of instalation of braze sdk: https://github.com/vitorreis/react-native-braze-sdk-app/commit/1692837480017f511708b96313ba45091de60c35

**Comment by radixdev** [comment]
This looks like it could be a build tools issue. See https://github.com/vitorreis/react-native-braze-sdk-app/blob/91fd6bba45280a6937803c6c4552b85446d1378c/android/app/build.gradle#L97-L103 

Can you try updating `compileSdkVersion`, `targetSdkVersion`, and `buildToolsVersion`?

**Comment by vitorreis** [comment]
What is the minimal version required by the SDK?

**Comment by Bucimis** [comment]
@vitorreis our React Native SDK requires API 16 and runs Braze Android SDK 2.4.0, see https://github.com/Appboy/appboy-react-sdk/blob/master/android/build.gradle. 

Could you also send us which versions of the Android support libraries you're using?

**Comment by radixdev** [comment]
I'd recommend using the latest values for each, so `28`, `28`, and `28.0.1`. We generally do not enforce a minimum version for the SDK, only a minimum as required for the support library to build the project.

**Comment by mattcolman** [comment]
I'm getting the same error. My build settings are:
```
android {
    compileSdkVersion 23
    buildToolsVersion "23.0.1"

    defaultConfig {
        minSdkVersion 16
        targetSdkVersion 22
        versionCode 1
        versionName "1.0"
```

**Comment by Bucimis** [comment]
@mattcolman what's your support library version?

**Comment by vitorreis** [comment]
@mattcolman I was able to solve this problem with an update of versions of android, see my `android/app/build.gradle` file:
```
android {
    compileSdkVersion 26
    buildToolsVersion "26.0.0"

    defaultConfig {
        applicationId "com.appwithbrazesdk"
        minSdkVersion 16
        targetSdkVersion 26
        versionCode 1
        versionName "1.0"
    }
}

dependencies {
    compile fileTree(dir: "libs", include: ["*.jar"])
    compile "com.android.support:appcompat-v7:26.0.0"
    compile "com.facebook.react:react-native:+"  // From node_modules
    compile project(':react-native-appboy-sdk')
}
```

I've updated sdk, buildtools and support library to v26.

In addition to that I had another problem, with `com.android.dex.DexException: Multiple dex files define` this happened on my case because of a conflict of appboy sdk with the wix react navigation library, I am still investigating how to solve this one.

**Comment by Bucimis** [comment]
Thanks @vitorreis - let us know how the `DexException` detective work goes

**Comment by mattcolman** [comment]
great thanks @vitorreis !

**Comment by vitorreis** [comment]
To explain the reasons of this error I will to give a bit of context.
I am working in react native app that was created a few weeks ago using the https://github.com/react-community/create-react-native-app package, I have ejected the app.

While attempting to implement the braze sdk in this app I got the errors reported in this thread and figure it out that I had to update the SDK compile version in order to get it to work, I've updated it and got a different compilation error and assumed it to be related to the braze sdk library, but turns out that it wasn't necessarily because of it. 

I tried then first to update the compile sdk version of the android project without adding the braze sdk and got the DexException as well, after a bit of investigation I concluded that it was caused by different versions of the android support library of my project and a dependency of [react-navigation](https://wix.github.io/react-native-navigation/#/) from wix that I'm using, this package depends on the `com.android.support:design` package.

I was able to solve the conflict with the guidance of this [article](https://github.com/transistorsoft/react-native-background-fetch/wiki/Solving-Android-Gradle-Conflicts) which btw could be applied to the appboy-android sdk as well (using gradle variables to dynamically set versions of dependencies).

After the conflict with react-navigation was solved by force overriding some dependencies I was able to setup the braze sdk.

Furthermore I think that the problem it's caused to some extent to the `eject` functionality of the `create-react-native-app` package, because it creates the android project with an outdated compile sdk version, this [article](https://medium.com/androiddevelopers/picking-your-compilesdkversion-minsdkversion-targetsdkversion-a098a0341ebd) states that it's a best practice always compile the android app with the latest sdk version, fact that was new to me since I'm not a mobile developer.
